captum/attr/_core/lime.py

```diff
@@ -0,0 +1,1007 @@
+                    expand_inputs = True
+
+            combined_interp_inps = torch.cat(interpretable_inps)
+            combined_outputs = (
+                torch.cat(outputs)
+                if len(outputs[0].shape) > 0
+                else torch.stack(outputs)
+            )
+            combined_sim = (
+                torch.cat(similarities)
+                if len(similarities[0].shape) > 0
+                else torch.stack(similarities)
+            )
+            if expand_inputs:
+                combined_interp_inps = torch.repeat_interleave(
+                    combined_interp_inps, bsz, 0
+                )
+                combined_sim = torch.repeat_interleave(combined_sim, bsz, 0)
+
+            interp_model = self.train_interpretable_model_func(
+                combined_interp_inps, combined_outputs, combined_sim, **kwargs
+            )
+            return interp_model
+
+    def has_convergence_delta(self) -> bool:
+        return False
+
+    @property
+    def multiplies_by_inputs(self):
+        return False
+
+
+def lasso_interpretable_model_trainer(
+    interp_inputs: Tensor, exp_outputs: Tensor, weights: Tensor, **kwargs
+):
+    try:
+        import sklearn
+        from sklearn import linear_model
+
+        assert (
+            sklearn.__version__ >= "0.23.0"
+        ), "Must have sklearn version 0.23.0 or higher to use "
+        "sample_weight in Lasso regression."
+    except ImportError:
+        raise AssertionError(
+            "Requires sklearn for default interpretable model training with"
+            " Lasso regression. Please install sklearn or use a custom interpretable"
+            " model training function."
+        )
+    clf = linear_model.Lasso(alpha=kwargs["alpha"] if "alpha" in kwargs else 1.0)
+    clf.fit(
+        interp_inputs.cpu().numpy(), exp_outputs.cpu().numpy(), weights.cpu().numpy()
+    )
+    return torch.from_numpy(clf.coef_)
+
+
+def default_from_interp_rep_transform(curr_sample, original_inputs, **kwargs):
+    assert (
+        "feature_mask" in kwargs
+    ), "Must provide feature_mask to use default interpretable representation transform"
+    assert (
+        "baselines" in kwargs
+    ), "Must provide baselines to use default interpretable representation transfrom"
+    feature_mask = kwargs["feature_mask"]
+    if isinstance(feature_mask, Tensor):
+        binary_mask = curr_sample[0][feature_mask]
+        return binary_mask * original_inputs + (1 - binary_mask) * kwargs["baselines"]
+    else:
+        binary_mask = tuple(
+            curr_sample[0][feature_mask[j]] for j in range(len(feature_mask))
+        )
+        return tuple(
+            binary_mask[j] * original_inputs[j]
+            + (1 - binary_mask[j]) * kwargs["baselines"][j]
+            for j in range(len(feature_mask))
+        )
+
+
+def default_similarity_kernel(original_inp, _, __, **kwargs):
+    return 1.0
```

Comment:
For lime don't we want to set similarity kernel to ?https://github.com/marcotcr/lime/blob/master/lime/lime_image.py#L121

Reply:
I was debating what to set the default; the [paper](https://arxiv.org/pdf/1602.04938.pdf) recommends an exponential kernel with distance function based on the use case (cosine distance for image, L2 for text, etc.). Supporting this exactly requires more argument for distance function type and kernel width, so I figured just a constant default would be generic, and a user can choose the appropriate kernel, distance function, and kernel width based on their use case. What do you think?

Reply:
I understand. That's the distance function, right? The kernel seems to be always defaulted to exponential kernel.
https://github.com/marcotcr/lime/blob/master/lime/lime_image.py#L121
https://github.com/marcotcr/lime/blob/master/lime/lime_tabular.py#L251
https://github.com/marcotcr/lime/blob/master/lime/lime_text.py#L353

Reply:
Correct, it's recommended that the distance is passed through an exponential kernel to convert to weights between 0 and 1, although this can also be varied to another kernel. We can default to applying an exponential kernel, but we'd have to add arguments for kernel width and distance function (or assume some fixed defaults for these as well), so this seemed simpler, instead of allowing the kernel, distance function, and width to all be set independently. I can add a couple of other pre-defined options for users to use e.g. exponential_kernel(dist_func, kernel_width), where a user provides a distance function and kernel width and constructs the appropriate similarity function?

Sorry, I think the naming for this default function may be confusing, I can update it to default_similarity_function to be consistent with the parameter name?

Reply:
Yeah, I think that at least providing `exponential_kernel ` and defaulting to it would be great because I assume that people usually don't define kernel function that often and we can default to exponential_kernel with fixed kernel_width and distance function would we good.

Reply:
It's interesting that for image they also defaulted to `cosine`. 
https://github.com/marcotcr/lime/blob/master/lime/lime_image.py#L134
https://github.com/marcotcr/lime/blob/master/lime/lime_image.py#L134

Both image and text are cosine and tabular Euclidean ?